lib/pg.ddl.js

```diff
@@ -0,0 +1,323 @@
+  pad('Id', length) + (type === 'Global' ? ' bigint' : ' bigserial');
+
+const generateTable = (
+  name,
+  table,
+  type,
+  types,
+  unresolved,
+  existingTables
+) => {
+  if (RESERVED.includes(name.toUpperCase())) {
+    throw new Error(
+      `Cannot create table ${name} because it is a reserved keyword`
+    );
+  }
+
+  const maxPropLength = getMaxPropLength(table);
+
+  const { indexes, unique, links, properties } = splitEntries(table, name);
+
+  const linksHere = unresolved.get(name) || [];
+  const unres = links.filter(({ name: propName, property: link }) => {
+    if (link.category === name || existingTables.includes(link.category)) {
+      linksHere.push({ from: name, to: link.category, name: propName, link });
+      return false;
+    }
+    return true;
+  });
+
+  const props = generateProperties(name, properties, types, maxPropLength);
+  props.unshift(generateId(type, maxPropLength));
+
+  const primaryKey =
+    `\nALTER TABLE ${name} ADD CONSTRAINT pk${name}Id PRIMARY KEY (Id);\n`;
+
+  let linksSQL = generateLinks(linksHere);
+  if (linksSQL) {
+    linksSQL = '\n' + linksSQL + '\n';
+  }
+
+  let indexesSQL = generateIndexes(indexes, name);
+  if (indexesSQL) {
+    indexesSQL = '\n' + indexesSQL + '\n';
+  }
+
+  let uniqueSQL = generateUnique(unique, name);
+  if (uniqueSQL) {
+    uniqueSQL = '\n' + uniqueSQL + '\n';
+  }
+
+  const sql =
+    `CREATE TABLE ${name} (\n  ${props.join(',\n  ')}\n);\n` +
+    indexesSQL +
+    uniqueSQL +
+    primaryKey +
+    linksSQL;
+
+  return {
+    sql,
+    unresolved: unres,
+  };
+};
+
+const tableTypes = {
+  Registry: 'Global',
+  Dictionary: 'Global',
+  System: 'Global',
+  History: 'Global',
+  Local: 'Local',
+  Log: 'Ignore',
+  View: 'Ignore',
+  Memory: 'Ignore',
+};
+
+const generateTables = (schema, types) => {
+  const unresolvedLinks = new Map(); /* links: table => [{ name, link }] */
+  const existingTables = [];
+  return iter(Object.keys(schema))
+    .map(name => {
+      const table = schema[name];
+      const type = tableTypes[table.constructor.name] || 'Local';
+      return {
+        name,
+        table,
+        type,
+      };
+    })
+    .filter(({ type }) => type !== 'Ignored')
+    .map(({ name, table, type }) => {
+      const { sql, unresolved } = generateTable(
+        name,
+        table,
+        type,
+        types,
+        unresolvedLinks,
+        existingTables
+      );
+      if (unresolved) {
+        unresolved.forEach(({ name: propName, property: link }) => {
+          const unresolved = unresolvedLinks.get(link.category);
+          const newLink = {
+            from: name,
+            to: link.category,
+            name: propName,
+            link,
+          };
+          if (unresolved) {
+            unresolved.push(newLink);
+          } else {
+            unresolvedLinks.set(link.category, [newLink]);
+          }
+        });
+      }
+      existingTables.push(name);
+      return sql;
+    })
+    .toArray()
```

Comment:
This way an array will be created only once (at the end), but if you use array methods every step will create an array so I'm -1 on this.

Reply:
@lundibundi I don't think performance is a concern for this part of the software, and I commented on that just because using the default array methods is more concise, I'm okay with it either way.